src/ir/LocalGraph.cpp

```diff
@@ -244,6 +244,43 @@ LocalGraph::LocalGraph(Function* func) {
 #endif
 }
 
+bool LocalGraph::equivalent(LocalGet* a, LocalGet* b) {
+  auto& aSets = getSetses[a];
+  auto& bSets = getSetses[b];
+  // The simple case of one set dominating two gets easily proves that they must
+  // have the same value. (Note that we can infer dominance from the fact that
+  // there is a single set: if the set did not dominate one of the gets then
+  // there would definitely be another set for that get, the zero initialization
+  // at the function entry, if nothing else.)
+  if (aSets.size() == 1 && bSets.size() == 1) {
+    auto* aSet = *aSets.begin();
+    auto* bSet = *bSets.begin();
+    if (aSet == bSet) {
+      if (!aSet) {
+        // They are both nullptr, indicating the implicit value for a parameter
+        // or the zero for a local.
+        if (func->isParam(a->index)) {
+          // For parameters to be equivalent they must have the exact same
+          // index.
+          return a->index == b->index;
+        } else {
+          // As locals, they are both of value zero, but must have the right
+          // type as well.
+          return func->getLocalType(a->index) == func->getLocalType(b->index);
+        }
+      } else {
+        // They are both the same actual set.
+        return true;
+      }
+    }
+    return false;
+  }
+  return false;
```

Comment:
This all would be easier to read as a sequence of early returns rather than as nested `if`s, IMO.